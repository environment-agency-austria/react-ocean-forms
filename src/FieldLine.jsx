/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';

import { fieldMetaShape, fieldShape } from './shapes';
import FieldError from './FieldError';
import FormText from './FormText';
import defaultValidators from './validators';

/**
 * Createa a required *
 * @param {Array} validators Validator array
 */
function createRequiredMarker(validators) {
  if (Array.isArray(validators) && validators.includes(defaultValidators.required)) {
    return <span className="field-required"> *</span>;
  }

  return null;
}

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
function FieldLine(props) {
  const {
    field,
    meta,
    label,
    children,
    validators,
  } = props;

  const groupClass = meta.valid ? 'field-group' : 'field-group is-invalid';

  /* eslint-disable jsx-a11y/label-has-for */
  return (
    <div className={groupClass}>
      <label htmlFor={field.id} className="text-right">
        <FormText text={label} />
        {createRequiredMarker(validators)}
      </label>
      <div className="input-container">
        {children}
        <FieldError
          id={`${field.id}_errors`}
          invalid={!meta.valid}
          error={meta.error}
          stringFormatter={meta.stringFormatter}
        />
      </div>
    </div>
  );
  /* eslint-enable jsx-a11y/label-has-for */
}

FieldLine.displayName = 'FieldLine';

FieldLine.defaultProps = {
  validators: undefined,
};

FieldLine.propTypes = {
  label: PropTypes.string.isRequired,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
};

export default FieldLine;
