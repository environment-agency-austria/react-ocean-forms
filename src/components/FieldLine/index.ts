export * from './FieldLine';
export * from './FieldLine.types';
