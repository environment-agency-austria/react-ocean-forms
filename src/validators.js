/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Wrapper function to call validators with parameters
 * @param {function} validator function to call
 * @param {object} context form context
 * @param {array} args parameters for the validator
 */
const withParam = (validator, ...args) => ((value, context) => validator(value, context, args));

/**
 * Checks if there is any value
 * @param {object} value field value
 */
const required = (value) => {
  // Special check for empty arrays
  if (Array.isArray(value) && value.length === 0) return 'ojs_error_required';
  if (value === 0) return undefined;

  return value ? undefined : 'ojs_error_required';
};

/**
 * Checks if the value is alpha numeric
 * TODO: Real implementation
 */
const alphaNumeric = value => (value && /[^a-zA-Z0-9 ]/i.test(value) ? 'ojs_error_alphaNumeric' : undefined);

/**
 * Checks if the given value has the minimum
 * length
 * @param {object} value field value
 * @param {object} context form context
 * @param {number} length minimum length
 */
const minLength = (value, context, [length]) => {
  if (value.length >= length) return undefined;

  return {
    message_id: 'ojs_error_minLength',
    params: {
      length: String(length),
    },
  };
};

/**
 * Checks if the given value has the maximum
 * length
 * @param {object} value field value
 * @param {object} context form context
 * @param {number} length maximum length
 */
const maxLength = (value, context, [length]) => {
  if (value.length <= length) return undefined;

  return {
    message_id: 'ojs_error_maxLength',
    params: {
      length: String(length),
    },
  };
};

export default {
  withParam,
  required,
  alphaNumeric,
  minLength,
  maxLength,
};
