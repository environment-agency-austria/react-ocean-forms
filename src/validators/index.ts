export * from './validators';
export * from './validators.types';
