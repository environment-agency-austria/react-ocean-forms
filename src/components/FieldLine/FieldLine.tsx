/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { FieldError } from '../FieldError';
import { FormText } from '../FormText';
import { IFieldLineProps } from './FieldLine.types';

/**
 * Create a required *
 * @param validators Validator array
 */
function createRequiredMarker(isRequired: boolean): JSX.Element | null {
  if (isRequired) {
    return <span className="field-required"> *</span>;
  }

  return null;
}

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
export const FieldLine: React.SFC<IFieldLineProps> = (props: IFieldLineProps): JSX.Element => {
  const {
    field,
    meta,
    label,
    children,
  } = props;

  const groupClass = meta.valid ? 'field-group' : 'field-group is-invalid';

  return (
    <div className={groupClass}>
      <label htmlFor={field.id} className="text-right">
        <FormText text={label} />
        {createRequiredMarker(meta.isRequired)}
      </label>
      <div className="input-container">
        {children}
        <FieldError
          id={`${field.id}_errors`}
          invalid={!meta.valid}
          error={meta.error}
          stringFormatter={meta.stringFormatter}
        />
      </div>
    </div>
  );
};

FieldLine.displayName = 'FieldLine';
