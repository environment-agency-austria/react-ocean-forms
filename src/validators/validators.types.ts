/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { IFormContext } from '../components';
import { IMessageValues } from '../utils/stringFormatter';

/**
 * Enum representing the message id's for field errors.
 */
export enum FieldErrorMessageId {
  AlphaNumeric = 'ojs_error_alphaNumeric',
  MaxLength = 'ojs_error_maxLength',
  MinLength = 'ojs_error_minLength',
  Required = 'ojs_error_required',
}

/**
 * Error object for formatting errors through
 * the stringFormatter method
 */
export interface IFieldErrorObject {
  /**
   * Message id of the error or raw
   * error string
   */
  message_id: string;
  /**
   * Optional parameters for the stringFormatter
   */
  params: IMessageValues;
}

/**
 * Returns true if the given object implements
 * IFIeldErrorObject
 * @param object Object to test
 */
export function isIFieldErrorObject(object: any): object is IFieldErrorObject {
  return object && typeof (<IFieldErrorObject>object).message_id === 'string';
}

/**
 * Possible return values of a validator
 */
export type TFieldError = undefined | string | IFieldErrorObject;
/**
 * Possible error states of a validated component
 */
export type TFieldErrors = null | IFieldErrorObject | IFieldErrorObject[];

/**
 * Validator method type
 */
export type TValidator<TFieldValue = unknown> = ((value: TFieldValue | undefined, context: IFormContext, ...args: unknown[]) => TFieldError);
/**
 * Async validator method type
 */
export type TAsyncValidator<TFieldValue = unknown> = ((value: TFieldValue | undefined, context: IFormContext, ...args: unknown[]) => Promise<TFieldError>);

/**
 * Default validator type
 */
export interface IDefaultValidator extends TValidator {
  isDefaultValidator: true;
}

/**
 * Returns true if the given object is a IDefaultValidator
 * @param object Function to test
 */
export function isDefaultValidator(object: any): object is IDefaultValidator {
  return object && typeof object === 'function' && (<IDefaultValidator>object).isDefaultValidator === true;
}
