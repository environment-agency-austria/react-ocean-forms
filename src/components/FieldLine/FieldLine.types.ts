import { TValidator } from '../../validators';
import { IFieldComponentProps } from '../Field';

/**
 * Field Line properties
 */
export interface IFieldLineProps extends IFieldComponentProps {
  /**
   * Field label
   */
  label: string;
  /**
   * Validators
   */
  validators?: TValidator[];
  /**
   * Children
   */
  children: React.ReactNode;
}
