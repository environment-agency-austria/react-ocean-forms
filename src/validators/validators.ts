/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { IFormContext } from '../components';
import { FieldErrorMessageId, TAsyncValidator, TFieldError, TValidator, IDefaultValidator } from './validators.types';

/**
 * Wrapper function to call validators with parameters
 * @param validator function to call
 * @param context form context
 * @param args parameters for the validator
 */
const withParam = (validator: TValidator, ...args: unknown[]): TValidator => {
  return (value: unknown, context: IFormContext): TFieldError => validator(value, context, args);
};
/**
 * Wrapper function to call async validators with parameters
 * @param validator function to call
 * @param context form context
 * @param args parameters for the validator
 */
const withAsyncParam = (validator: TAsyncValidator, ...args: unknown[]): TAsyncValidator => {
  return async (value: unknown, context: IFormContext): Promise<TFieldError> => validator(value, context, args);
};

/**
 * Checks if there is any value
 * @param value field value
 */
const required: IDefaultValidator = (value: unknown): TFieldError => {
  // Special check for empty arrays
  if (Array.isArray(value)) {
    return value.length === 0 ? FieldErrorMessageId.Required : undefined;
  }

  if (typeof value === 'number') {
    return undefined;
  }

  if (typeof value === 'boolean') {
    return value === false ? FieldErrorMessageId.Required : undefined;
  }

  if (typeof value === 'string') {
    return value.length === 0 ? FieldErrorMessageId.Required : undefined;
  }

  return value !== null && value !== undefined ? undefined : FieldErrorMessageId.Required;
};
required.isDefaultValidator = true as const;

/**
 * Checks if the value is alpha numeric
 */
const alphaNumeric = (value: unknown): TFieldError => {
  if (typeof value !== 'string') { return undefined; }

  return /[^a-zA-Z0-9 ]/i.test(value) ? FieldErrorMessageId.AlphaNumeric : undefined;
};

interface ILength {
  length: number;
}

function isILength(object: any): object is ILength {
  return object !== null && object !== undefined && typeof (<ILength>object).length === 'number';
}

/**
 * Checks if the given value has the minimum
 * length
 * @param value field value
 * @param context form context
 * @param length minimum length
 */
const minLength = (value: unknown, context: IFormContext, [length]: [number]): TFieldError => {
  if (!isILength(value)) { return undefined; }
  if (value.length >= length) { return undefined; }

  return {
    message_id: FieldErrorMessageId.MinLength,
    params: {
      length: String(length),
    },
  };
};

/**
 * Checks if the given value has the maximum
 * length
 * @param value field value
 * @param context form context
 * @param length maximum length
 */
const maxLength = (value: unknown, context: IFormContext, [length]: [number]): TFieldError => {
  if (!isILength(value)) { return undefined; }
  if (value.length <= length) { return undefined; }

  return {
    message_id: FieldErrorMessageId.MaxLength,
    params: {
      length: String(length),
    },
  };
};

export const validators = {
  withParam,
  withAsyncParam,
  required,
  alphaNumeric,
  minLength,
  maxLength,
};
