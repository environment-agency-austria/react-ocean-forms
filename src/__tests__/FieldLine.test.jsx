import React from 'react';
import { shallow } from 'enzyme';

import validators from '../validators';
import FieldLine from '../FieldLine';

describe('<FieldLine />', () => {
  const meta = {
    valid: true,
    error: null,
    isValidating: false,
    touched: false,
    stringFormatter: jest.fn(),
  };
  const field = {
    id: 'unitInput',
    name: 'unitInput',
    value: '',
    disabled: false,
    onChange: jest.fn(),
    onBlur: jest.fn(),
  };

  const setup = props => shallow((
    <FieldLine
      id="unitInput"
      label="unitLabel"
      meta={meta}
      field={field}
      {...props}
    >
      <div>children</div>
    </FieldLine>
  ));

  it('should render without crashing', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should display invalid correctly', () => {
    meta.valid = false;
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should show a required marker', () => {
    const wrapper = setup({
      validators: [validators.required],
    });
    expect(wrapper).toMatchSnapshot();
  });
});
