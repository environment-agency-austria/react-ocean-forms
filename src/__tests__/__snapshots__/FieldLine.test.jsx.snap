// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldLine /> should display invalid correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FieldLine
    field={
      Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    id="unitInput"
    label="unitLabel"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "stringFormatter": [MockFunction] {
          "calls": Array [
            Array [
              "unitLabel",
            ],
            Array [
              "unitLabel",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "touched": false,
        "valid": false,
      }
    }
  >
    <div>
      children
    </div>
  </FieldLine>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          className="text-right"
          htmlFor="unitInput"
        />,
        <div
          className="input-container"
        >
          <div>
            children
          </div>
          <FieldError
            error={null}
            id="unitInput_errors"
            invalid={true}
            stringFormatter={
              [MockFunction] {
                "calls": Array [
                  Array [
                    "unitLabel",
                  ],
                  Array [
                    "unitLabel",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
          />
        </div>,
      ],
      "className": "field-group is-invalid",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            null,
          ],
          "className": "text-right",
          "htmlFor": "unitInput",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          null,
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div>
              children
            </div>,
            <FieldError
              error={null}
              id="unitInput_errors"
              invalid={true}
              stringFormatter={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      "unitLabel",
                    ],
                    Array [
                      "unitLabel",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
            />,
          ],
          "className": "input-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "children",
            },
            "ref": null,
            "rendered": "children",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "error": null,
              "id": "unitInput_errors",
              "invalid": true,
              "stringFormatter": [MockFunction] {
                "calls": Array [
                  Array [
                    "unitLabel",
                  ],
                  Array [
                    "unitLabel",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            className="text-right"
            htmlFor="unitInput"
          />,
          <div
            className="input-container"
          >
            <div>
              children
            </div>
            <FieldError
              error={null}
              id="unitInput_errors"
              invalid={true}
              stringFormatter={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      "unitLabel",
                    ],
                    Array [
                      "unitLabel",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
            />
          </div>,
        ],
        "className": "field-group is-invalid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              null,
            ],
            "className": "text-right",
            "htmlFor": "unitInput",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            null,
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div>
                children
              </div>,
              <FieldError
                error={null}
                id="unitInput_errors"
                invalid={true}
                stringFormatter={
                  [MockFunction] {
                    "calls": Array [
                      Array [
                        "unitLabel",
                      ],
                      Array [
                        "unitLabel",
                      ],
                    ],
                    "results": Array [
                      Object {
                        "isThrow": false,
                        "value": undefined,
                      },
                      Object {
                        "isThrow": false,
                        "value": undefined,
                      },
                    ],
                  }
                }
              />,
            ],
            "className": "input-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "children",
              },
              "ref": null,
              "rendered": "children",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "error": null,
                "id": "unitInput_errors",
                "invalid": true,
                "stringFormatter": [MockFunction] {
                  "calls": Array [
                    Array [
                      "unitLabel",
                    ],
                    Array [
                      "unitLabel",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<FieldLine /> should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FieldLine
    field={
      Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    id="unitInput"
    label="unitLabel"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "stringFormatter": [MockFunction] {
          "calls": Array [
            Array [
              "unitLabel",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "touched": false,
        "valid": true,
      }
    }
  >
    <div>
      children
    </div>
  </FieldLine>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          className="text-right"
          htmlFor="unitInput"
        />,
        <div
          className="input-container"
        >
          <div>
            children
          </div>
          <FieldError
            error={null}
            id="unitInput_errors"
            invalid={false}
            stringFormatter={
              [MockFunction] {
                "calls": Array [
                  Array [
                    "unitLabel",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
          />
        </div>,
      ],
      "className": "field-group",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            null,
          ],
          "className": "text-right",
          "htmlFor": "unitInput",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          null,
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div>
              children
            </div>,
            <FieldError
              error={null}
              id="unitInput_errors"
              invalid={false}
              stringFormatter={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      "unitLabel",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
            />,
          ],
          "className": "input-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "children",
            },
            "ref": null,
            "rendered": "children",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "error": null,
              "id": "unitInput_errors",
              "invalid": false,
              "stringFormatter": [MockFunction] {
                "calls": Array [
                  Array [
                    "unitLabel",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            className="text-right"
            htmlFor="unitInput"
          />,
          <div
            className="input-container"
          >
            <div>
              children
            </div>
            <FieldError
              error={null}
              id="unitInput_errors"
              invalid={false}
              stringFormatter={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      "unitLabel",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
            />
          </div>,
        ],
        "className": "field-group",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              null,
            ],
            "className": "text-right",
            "htmlFor": "unitInput",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            null,
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div>
                children
              </div>,
              <FieldError
                error={null}
                id="unitInput_errors"
                invalid={false}
                stringFormatter={
                  [MockFunction] {
                    "calls": Array [
                      Array [
                        "unitLabel",
                      ],
                    ],
                    "results": Array [
                      Object {
                        "isThrow": false,
                        "value": undefined,
                      },
                    ],
                  }
                }
              />,
            ],
            "className": "input-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "children",
              },
              "ref": null,
              "rendered": "children",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "error": null,
                "id": "unitInput_errors",
                "invalid": false,
                "stringFormatter": [MockFunction] {
                  "calls": Array [
                    Array [
                      "unitLabel",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<FieldLine /> should show a required marker 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FieldLine
    field={
      Object {
        "disabled": false,
        "id": "unitInput",
        "name": "unitInput",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    id="unitInput"
    label="unitLabel"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "stringFormatter": [MockFunction] {
          "calls": Array [
            Array [
              "unitLabel",
            ],
            Array [
              "unitLabel",
            ],
            Array [
              "unitLabel",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "touched": false,
        "valid": false,
      }
    }
    validators={
      Array [
        [Function],
      ]
    }
  >
    <div>
      children
    </div>
  </FieldLine>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          className="text-right"
          htmlFor="unitInput"
        >
          <span
            className="field-required"
          >
             *
          </span>
        </label>,
        <div
          className="input-container"
        >
          <div>
            children
          </div>
          <FieldError
            error={null}
            id="unitInput_errors"
            invalid={true}
            stringFormatter={
              [MockFunction] {
                "calls": Array [
                  Array [
                    "unitLabel",
                  ],
                  Array [
                    "unitLabel",
                  ],
                  Array [
                    "unitLabel",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
          />
        </div>,
      ],
      "className": "field-group is-invalid",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            <span
              className="field-required"
            >
               *
            </span>,
          ],
          "className": "text-right",
          "htmlFor": "unitInput",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " *",
              "className": "field-required",
            },
            "ref": null,
            "rendered": " *",
            "type": "span",
          },
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div>
              children
            </div>,
            <FieldError
              error={null}
              id="unitInput_errors"
              invalid={true}
              stringFormatter={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      "unitLabel",
                    ],
                    Array [
                      "unitLabel",
                    ],
                    Array [
                      "unitLabel",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
            />,
          ],
          "className": "input-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "children",
            },
            "ref": null,
            "rendered": "children",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "error": null,
              "id": "unitInput_errors",
              "invalid": true,
              "stringFormatter": [MockFunction] {
                "calls": Array [
                  Array [
                    "unitLabel",
                  ],
                  Array [
                    "unitLabel",
                  ],
                  Array [
                    "unitLabel",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            className="text-right"
            htmlFor="unitInput"
          >
            <span
              className="field-required"
            >
               *
            </span>
          </label>,
          <div
            className="input-container"
          >
            <div>
              children
            </div>
            <FieldError
              error={null}
              id="unitInput_errors"
              invalid={true}
              stringFormatter={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      "unitLabel",
                    ],
                    Array [
                      "unitLabel",
                    ],
                    Array [
                      "unitLabel",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
            />
          </div>,
        ],
        "className": "field-group is-invalid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              <span
                className="field-required"
              >
                 *
              </span>,
            ],
            "className": "text-right",
            "htmlFor": "unitInput",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " *",
                "className": "field-required",
              },
              "ref": null,
              "rendered": " *",
              "type": "span",
            },
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div>
                children
              </div>,
              <FieldError
                error={null}
                id="unitInput_errors"
                invalid={true}
                stringFormatter={
                  [MockFunction] {
                    "calls": Array [
                      Array [
                        "unitLabel",
                      ],
                      Array [
                        "unitLabel",
                      ],
                      Array [
                        "unitLabel",
                      ],
                    ],
                    "results": Array [
                      Object {
                        "isThrow": false,
                        "value": undefined,
                      },
                      Object {
                        "isThrow": false,
                        "value": undefined,
                      },
                      Object {
                        "isThrow": false,
                        "value": undefined,
                      },
                    ],
                  }
                }
              />,
            ],
            "className": "input-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "children",
              },
              "ref": null,
              "rendered": "children",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "error": null,
                "id": "unitInput_errors",
                "invalid": true,
                "stringFormatter": [MockFunction] {
                  "calls": Array [
                    Array [
                      "unitLabel",
                    ],
                    Array [
                      "unitLabel",
                    ],
                    Array [
                      "unitLabel",
                    ],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
